man/man8/puppetdoc.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PUPPET\-DOC" "8" "June 2012" "Puppet Labs, LLC" "Puppet manual"
.
.SH "NAME"
\fBpuppet\-doc\fR \- Generate Puppet documentation and references
.
.SH "SYNOPSIS"
Generates a reference for all Puppet types\. Largely meant for internal Puppet Labs use\.
.
.SH "USAGE"
puppet doc [\-a|\-\-all] [\-h|\-\-help] [\-o|\-\-outputdir \fIrdoc\-outputdir\fR] [\-m|\-\-mode text|pdf|rdoc] [\-r|\-\-reference \fIreference\-name\fR] [\-\-charset \fIcharset\fR] [\fImanifest\-file\fR]
.
.SH "DESCRIPTION"
If mode is not \'rdoc\', then this command generates a Markdown document describing all installed Puppet types or all allowable arguments to puppet executables\. It is largely meant for internal use and is used to generate the reference document available on the Puppet Labs web site\.
.
.P
In \'rdoc\' mode, this command generates an html RDoc hierarchy describing the manifests that are in \'manifestdir\' and \'modulepath\' configuration directives\. The generated documentation directory is doc by default but can be changed with the \'outputdir\' option\.
.
.P
If the command is run with the name of a manifest file as an argument, puppet doc will output a single manifest\'s documentation on stdout\.
.
.SH "OPTIONS"
.
.TP
\-\-all
Output the docs for all of the reference types\. In \'rdoc\' mode, this also outputs documentation for all resources\.
.
.TP
\-\-help
Print this help message
.
.TP
\-\-outputdir
Used only in \'rdoc\' mode\. The directory to which the rdoc output should be written\.
.
.TP
\-\-mode
Determine the output mode\. Valid modes are \'text\', \'pdf\' and \'rdoc\'\. The \'pdf\' mode creates PDF formatted files in the /tmp directory\. The default mode is \'text\'\. In \'rdoc\' mode you must provide \'manifests\-path\'
.
.TP
\-\-reference
Build a particular reference\. Get a list of references by running \'puppet doc \-\-list\'\.
.
.TP
\-\-charset
Used only in \'rdoc\' mode\. It sets the charset used in the html files produced\.
.
.TP
\-\-manifestdir
Used only in \'rdoc\' mode\. The directory to scan for stand\-alone manifests\. If not supplied, puppet doc will use the manifestdir from puppet\.conf\.
.
.TP
\-\-modulepath
Used only in \'rdoc\' mode\. The directory or directories to scan for modules\. If not supplied, puppet doc will use the modulepath from puppet\.conf\.
.
.TP
\-\-environment
Used only in \'rdoc\' mode\. The configuration environment from which to read the modulepath and manifestdir settings, when reading said settings from puppet\.conf\. Due to a known bug, this option is not currently effective\.
.
.SH "EXAMPLE"
.
.nf

$ puppet doc \-r type > /tmp/type_reference\.markdown
.
.fi
.
.P
or
.
.IP "" 4
.
.nf

$ puppet doc \-\-outputdir /tmp/rdoc \-\-mode rdoc /path/to/manifests
.
.fi
.
.IP "" 0
.
.P
or
.
.IP "" 4
.
.nf

$ puppet doc /etc/puppet/manifests/site\.pp
.
.fi
.
.IP "" 0
.
.P
or
.
.IP "" 4
.
.nf

$ puppet doc \-m pdf \-r configuration
.
.fi
.
.IP "" 0
.
.SH "AUTHOR"
Luke Kanies
.
.SH "COPYRIGHT"
Copyright (c) 2011 Puppet Labs, LLC Licensed under the Apache 2\.0 License


CHANGELOG
3.0.0rc3
===
e619304 (#14761) Add boot, reboot to excludes list for redhat provider
7c1d8c2 (maint) Disable upstart spec test on windows
37df5c4 (#14615) Final stub fixes for the tests.
f11dfd6 (#14615) Fix provider spec tests to work on non gentoo systems
060810e maint: Make it easier to test the service provider
c0cb0fa (Maint) Close filehandles after read in upstart
55505db (Maint) Fix upstart provider to work on Ruby 1.9.3
b0f1d58 Update redhat.spec dependency to include hiera-puppet
6e2f23d Update puppet gem dependencies
d19f93f Update redhat init scripts to use `puppet (agent|queue|master)`
a45942b (#14297) Don't remove tty1 or rsyslog during upstart acceptance
466ab49 (#14297) Define agent in upstart acceptance test
f5d1f74 (#14297) Update upstart provider acceptance tests
3c5ef40 (#14297) Clean up upstart provider and tests
36c41e1 (#14343) Test cases for more forms of variable lookup
6cca061 (#14297) Add acceptance test for upstart provider
b9ed026 (#14297) Update spec tests for upstart provider
91827e8 (#14297) Handle upstart services better
3541db6 (#14343) Lookup in correct inherited scope
2d83e22 (Maint) Added test for fully qualified variables
0788ea2 (#14615) Exclude helperscripts in gentoo service provider
2864d86 (maint) Change Facter dep from exit to raise
2ec9b74 (maint) Make Puppet have hard dependency on Facter 2.0
6c37538 (#3909) Move default certname logic to Puppet::Settings
6be09f0 (#3909) Strip trailing dot from default FQDN

3.0.0rc2
===
0cea47e (#14609) Fix master initialization when running under Passenger
b6c5e9d (#14514) Check for the presence of hiera config
ca5271f (maint) Add data binding Hiera tests

3.0.0rc1
===
a731e15 Update CHANGELOG lib/puppet.rb puppet.spec for 2.7.15rc1
53e0057 Use internal domain name for module tool tests
f4ba59e Use absolute_path in module install spec
ead9d25 (Maint) Color console output when the win32console gem is installed
a76d873 (#14424) Expand path of the target directory
95b75e1 Fix unit test failure in fileserver due to modules
da5361d Maint: De-genericize Puppet Forge references in module face
3b144bf Revert "maint: use relative paths to fixtures."
dae5539 (#14440) Hack to allow help faces to be used as API
e862cad Fix broken man page rake task
b26699a (#10146) `-` is not legal in variable names.
7ce9e17 maint: use relative paths to fixtures.
a7a532c Significantly improve compilation performance when using modules
12c5bda Fix a stub of ansicolor that I missed
ea740f6 Add Puppet::Util::Platform to abstract platform checks
91a4a7d Don't use features to check for color in logging
80fe59d Remove a bad test that can lead to random hangs...
0efb4c2 (#14296) Improve template function error message.
d5e0cdd (#14288) gem provider should tolerate bad lines.
574e4c8 Update CHANGELOG, spec files for 2.7.14
d2b3ba0 Revert "Fix incorrect argument handling for expire in NodeExpirer"
7c7d4b6 Fix #14123 for Windows
1c9bec0 Fix incorrect argument handling for expire in NodeExpirer
7e92a7c Fix filebucket specs on Windows
a713f07 (#8778) Make '' == undef commutative in the DSL
24d4bb5 Revert "(#11004) Fix regex for pkg provider line matching"
34da6f7 (#13651) Use magenta, not purple in Facter::Util::Colors
e37b536 (#11004) Fix regex for pkg provider line matching
9b47110 (#13956) Fix failing spec tests
7e92dcb Just change Puppet settings in tests, rather than stubbing.
ca313dc (maint) Fix destinations_spec failure on windows
cde3bee Just change Puppet settings in tests, rather than stubbing.
6aa8441 (#11004) Support Solaris 11 output in pkg package provider
61e7d3c maint: Fix pkg_spec test
603b368 (#14173) Enforce that filebucket paths must be absolute
b7828ed (#3452) Autorequire user for cron
a47c42d Update CHANGELOG for 2.7.14rc3
402a425 (#14127) Add integration tests for ssh_authorized_key
85ee441 maint: refactor integration specs for ssh_authorized_key
42b20be (#13682) Fix acceptance test failures
dce3c32 (#13682) Rename Puppet::Module::Tool to Puppet::ModuleTool
5117e12 (maint) Ensure every file has a trailing newlines
b4d1c65 (#14127) ssh_authorized_keys grammer fails on blank lines.
25ad6d7 Updating CHANGELOG for Puppet 2.7.14rc2
74e462f (#14060) Fix quoting of commands to interpolate inside the shell.
796a072 Modernize the style of the shell exec provider tests.
9e526c9 (#14060) Fix quoting of commands to interpolate inside the shell.
36d959a (#14123) Puppet shouldn't explode if PATH contains ~nonexistent_user
a009c54 Modernize the style of the shell exec provider tests.
934ec89 Improve deprecation warning for dynamic lookup
c255414 (#14036) Add upstart tests
dc3bb9f (#14036) Handle upstart better
a281c48 (#13645) Add test cases to tagmail_spec
3a4ac60 Updating CHANGELOG, lib/puppet.rb for 2.7.14rc1
f6ac1bb (#13645) Do not open a smtp connection if nothing to report
f40cc71 (#12392): Colorize console output on Windows
0d9e852 Removed duplication of compiling a catalog
d67711d Cleaner test for create_resources doing nothing
2d3cc63 Fixing problem caused by activerecord
f1331ba Put back require 'puppet' in Puppet::Util
79bf4fa (#13299) Fix error message when agent is disabled
5178adc Update puppet.spec for rpm packaging
fb96747 Fixed old log test to match new autoflush behavior
878127f Make sure the log file writable
72d3b56 Default autoflushing of log files to true
9711d50 Created test that shows enc warning problem
d54100d (Maint) Fix order-dependent test failure
f531400 (Maint) Don't assume eventlog gem is installed on Windows
2957ffe Fix tempfile cleanup in specs
e2d6915 Fix communication_error acceptance test
27a5eb7 options[:ca_location] is a string not a symbol. Safer to pull Puppet::SSL::Host.ca_location since it is already set correctly.
c01cac9 (#13204) Don't ignore missing PATH.augnew files
7d27cd2 Use :as_platform in the smf service spec
876785a Use :as_platform in the redhat service spec

2.7.16rc1
===
3a06246 (#8858) Explicitly set SSL peer verification mode.
cd4bee8 Test SSL setup code, not our stubbing of it.

lib/puppet/application/puppetdoc.rb
require 'puppet'
require 'puppet/application'
require 'puppet/util/reference'
require 'puppet/network/handler'
require 'puppet/util/rdoc'

$tab = "    "
Reference = Puppet::Util::Reference

Puppet::Application.new(:puppetdoc) do

    should_not_parse_config

    attr_accessor :unknown_args, :manifest

    preinit do
        {:references => [], :mode => :text, :format => :to_rest }.each do |name,value|
            options[name] = value
        end
        @unknown_args = []
        @manifest = false
    end

    option("--all","-a")
    option("--outputdir OUTPUTDIR","-o")
    option("--verbose","-v")
    option("--debug","-d")

    option("--format FORMAT", "-f") do |arg|
        method = "to_%s" % arg
        if Reference.method_defined?(method)
            options[:format] = method
        else
            raise "Invalid output format %s" % arg
        end
    end

    option("--mode MODE", "-m") do |arg|
        if Reference.modes.include?(arg) or arg.intern==:rdoc
            options[:mode] = arg.intern
        else
            raise "Invalid output mode %s" % arg
        end
    end

    option("--list", "-l") do |arg|
        puts Reference.references.collect { |r| Reference.reference(r).doc }.join("\n")
        exit(0)
    end

    option("--reference REFERENCE", "-r") do |arg|
        options[:references] << arg.intern
    end

    unknown do |opt, arg|
        @unknown_args << {:opt => opt, :arg => arg }
        true
    end

    dispatch do
        return options[:mode] if [:rdoc, :trac, :markdown].include?(options[:mode])
        return :other
    end

    command(:rdoc) do
        exit_code = 0
        files = []
        unless @manifest
            env = Puppet::Node::Environment.new
            files += env.modulepath
            files << File.dirname(env[:manifest])
        end
        files += ARGV
        Puppet.info "scanning: %s" % files.inspect
        Puppet.settings.setdefaults("puppetdoc",
            "document_all" => [false, "Document all resources"]
        )
        Puppet.settings[:document_all] = options[:all] || false
        begin
            if @manifest
                Puppet::Util::RDoc.manifestdoc(files)
            else
                options[:outputdir] = "doc" unless options[:outputdir]
                Puppet::Util::RDoc.rdoc(options[:outputdir], files)
            end
        rescue => detail
            if Puppet[:trace]
                puts detail.backtrace
            end
            $stderr.puts "Could not generate documentation: %s" % detail
            exit_code = 1
        end
        exit exit_code
    end

    command(:trac) do
        options[:references].each do |name|
            section = Puppet::Util::Reference.reference(name) or raise "Could not find section %s" % name
            unless options[:mode] == :pdf
                section.trac
            end
        end
    end

    command(:markdown) do
        text = ""
        with_contents = false
        exit_code = 0
        options[:references].sort { |a,b| a.to_s <=> b.to_s }.each do |name|
            raise "Could not find reference %s" % name unless section = Puppet::Util::Reference.reference(name)

            begin
                # Add the per-section text, but with no ToC
                text += section.send(options[:format], with_contents)
                text += Puppet::Util::Reference.footer
                text.gsub!(/`\w+\s+([^`]+)`:trac:/) { |m| $1 }
                Puppet::Util::Reference.markdown(name, text)
                text = ""
            rescue => detail
                puts detail.backtrace
                $stderr.puts "Could not generate reference %s: %s" % [name, detail]
                exit_code = 1
                next
            end
        end

        exit exit_code
    end

    command(:other) do
        text = ""
        if options[:references].length > 1
            with_contents = false
        else
            with_contents = true
        end
        exit_code = 0
        options[:references].sort { |a,b| a.to_s <=> b.to_s }.each do |name|
            raise "Could not find reference %s" % name unless section = Puppet::Util::Reference.reference(name)

            begin
                # Add the per-section text, but with no ToC
                text += section.send(options[:format], with_contents)
            rescue => detail
                puts detail.backtrace
                $stderr.puts "Could not generate reference %s: %s" % [name, detail]
                exit_code = 1
                next
            end
        end

        unless with_contents # We've only got one reference
            text += Puppet::Util::Reference.footer
        end

        # Replace the trac links, since they're invalid everywhere else
        text.gsub!(/`\w+\s+([^`]+)`:trac:/) { |m| $1 }

        if options[:mode] == :pdf
            Puppet::Util::Reference.pdf(text)
        else 
            puts text
        end

        exit exit_code
    end

    setup do
        # sole manifest documentation
        if ARGV.size > 0
            options[:mode] = :rdoc
            @manifest = true
        end

        if options[:mode] == :rdoc
            setup_rdoc
        else
            setup_reference
        end
    end

    def setup_reference
        if options[:all]
            # Don't add dynamic references to the "all" list.
            options[:references] = Reference.references.reject do |ref|
                Reference.reference(ref).dynamic?
            end
        end

        if options[:references].empty?
            options[:references] << :type
        end
    end

    def setup_rdoc(dummy_argument=:work_arround_for_ruby_GC_bug)
        # consume the unknown options
        # and feed them as settings
        if @unknown_args.size > 0
            @unknown_args.each do |option|
                # force absolute path for modulepath when passed on commandline
                if option[:opt]=="--modulepath" or option[:opt] == "--manifestdir"
                    option[:arg] = option[:arg].split(':').collect { |p| File.expand_path(p) }.join(':')
                end
                Puppet.settings.handlearg(option[:opt], option[:arg])
            end
        end

        # hack to get access to puppetmasterd modulepath and manifestdir
        Puppet[:name] = "puppetmasterd"
        # Now parse the config
        Puppet.parse_config

        # Handle the logging settings.
        if options[:debug] or options[:verbose]
            if options[:debug]
                Puppet::Util::Log.level = :debug
            else
                Puppet::Util::Log.level = :info
            end

            Puppet::Util::Log.newdestination(:console)
        end
    end
end
